README.md
# react-native-switch-toggle

> Simple switch toggle component for react-native. This component supports horizontal switch toggle with animation with several options like start/end background colors, start/end circle colors, and duration for animation.

[![Npm Version](http://img.shields.io/npm/v/react-native-switch-toggle.svg?style=flat-square)](https://npmjs.org/package/react-native-switch-toggle)
[![Downloads](http://img.shields.io/npm/dm/react-native-switch-toggle.svg?style=flat-square)](https://npmjs.org/package/react-native-switch-toggle)
[![CI](https://github.com/hyochan/react-native-switch-toggle/actions/workflows/ci.yml/badge.svg)](https://github.com/hyochan/react-native-switch-toggle/actions/workflows/ci.yml)
[![codecov](https://codecov.io/gh/hyochan/react-native-switch-toggle/branch/master/graph/badge.svg?token=yLFcpGSuku)](https://codecov.io/gh/hyochan/react-native-switch-toggle)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square)](https://github.com/prettier/prettier)
[![LICENSE](http://img.shields.io/npm/l/react-native-switch-toggle.svg?style=flat-square)](https://npmjs.org/package/react-native-switch-toggle)
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square)](CONTRIBUTING.md)
[![supports iOS](https://img.shields.io/badge/iOS-4630EB.svg?style=flat-square&logo=APPLE&labelColor=999999&logoColor=fff)](https://itunes.apple.com/app/apple-store/id982107779)
[![supports Android](https://img.shields.io/badge/Android-4630EB.svg?style=flat-square&logo=ANDROID&labelColor=A4C639&logoColor=fff)](https://play.google.com/store/apps/details?id=host.exp.exponent&referrer=www)
[![supports web](https://img.shields.io/badge/web-4630EB.svg?style=flat-square&logo=GOOGLE-CHROME&labelColor=4285F4&logoColor=fff)](https://docs.expo.io/workflow/web/)
[![runs with expo](https://img.shields.io/badge/Runs%20with%20Expo-000.svg?style=flat&logo=EXPO&labelColor=ffffff&logoColor=000)](https://github.com/expo/expo)


<img src="https://user-images.githubusercontent.com/27461460/120913849-32f23580-c6d5-11eb-90bc-be0a97125427.gif" height=400 />

## News

Current package is fully redesigned in `2.0.0`. We recommend you to use this package version `>=2.0.0`.

## Alternative

Do you enjoy using `react-native-switch-toggle`? Then think about using [dooboo-ui](https://github.com/dooboolab/dooboo-ui) which is our ui integrated ilbrary. It included `SwitchToggle` from `0.1.6`.

```
yarn add doobo-ui
```

## Props

|                      | necessary | types   | default                                                  |
| -------------------- | --------- | ------- | -------------------------------------------------------- |
| switchOn             | ✓         | boolean | false                                                    |
| onPress              | ✓         | func    | () => {}                                                 |
| containerStyle       |           | styles  | { width: 80, height: 40, borderRadius: 25 ... }          |
| circleStyle          |           | styles  | { width: 32, height: 32, borderRadius: 16 ... }          |
| backgroundColorOn    |           | string  | 'black'   |
| backgroundColorOff   |           | string  | '#C4C4C4' |
| circleColorOn        |           | string  | 'white'   |
| circleColorOff       |           | string  | '#C4C4C4' |
| duration             |           | number  | 300       |
| type                 |           | number  | 0 - Normal switch, 1 - Switch with a text                |
| buttonText           |           | string  | Text on-top of the button                                |
| backTextRight        |           | string  | Text to appear in right side when button toggled to left |
| backTextLeft         |           | string  | Text to appear in left side when button toggled to right |
| buttonTextStyle      |           | styles  | Styles for text on the button                            |
| textRightStyle       |           | styles  | Styles for text in right                                 |
| textLeftStyle        |           | styles  | Styles for text in left                                  |
| buttonStyle          |           | styles  | Styles for the button                                    |
| buttonContainerStyle |           | styles  | Styles for button container                              |
| rightContainerStyle  |           | styles  | Styles for right text container                          |
| leftContainerStyle   |           | styles  | Styles for left text container                           |

## Getting started

`$ npm install react-native-switch-toggle --save`

```javascript
import SwitchToggle from "react-native-switch-toggle";
```

#### Basic Style
```jsx
<SwitchToggle switchOn={on} onPress={() => off(!on)} />
```

#### Custom Color
```jsx
<SwitchToggle
  switchOn={on}
  onPress={() => off(!on)}
  circleColorOff='#C4C4C4'
  circleColorOn='#00D9D5'
  backgroundColorOn='#6D6D6D'
  backgroundColorOff='#C4C4C4'
/>
```

### Custom Size
```jsx
<SwitchToggle
  switchOn={on}
  onPress={() => off(!on)}
  containerStyle={{
    marginTop: 16,
    width: 106,
    height: 48,
    borderRadius: 25,
    padding: 5,
  }}
  circleStyle={{
    width: 40,
    height: 40,
    borderRadius: 20,
  }}
/>
```


CHANGELOG.md
[2.2.1]
Fix CI/CD to deploy `typescript` and `flow`.

[2.2.0]
Fix typescript not being exposed during deploy. [#49](https://github.com/hyochan/react-native-switch-toggle/issues/49)

Fix transition `inputRange` start from `0` [#48](https://github.com/hyochan/react-native-switch-toggle/issues/48).

[2.0.1]
Update readme.

[2.0.0]
Upgrade reliability and maintenance structures.


index.js
import React, { Component } from 'react';

import { Animated, StyleSheet, Text, TouchableOpacity } from 'react-native';

import PropTypes from 'prop-types';

class Item extends Component {
  static defaultProps = {
    switchOn: false,
    onPress: () => {},
    containerStyle: {
      width: 72,
      height: 36,
      borderRadius: 18,
      backgroundColor: 'rgb(227,227,227)',
      padding: 3,
    },
    circleStyle: {
      width: 30,
      height: 30,
      borderRadius: 15,
      backgroundColor: 'white', // rgb(102,134,205)
    },
    backgroundColorOn: 'rgb(227,227,227)',
    backgroundColorOff: 'rgb(215,215,215)',
    circleColorOff: 'white',
    circleColorOn: 'rgb(102,134,205)',
    duration: 300,
  };

  static propTypes = {
    switchOn: PropTypes.bool,
    onPress: PropTypes.func,
    containerStyle: PropTypes.any,
    circleStyle: PropTypes.any,
    backgroundColorOff: PropTypes.string,
    backgroundColorOn: PropTypes.string,
    circleColorOff: PropTypes.string,
    circleColorOn: PropTypes.string,
    duration: PropTypes.number,
    type: PropTypes.number,

    buttonText: PropTypes.string,
    backTextRight: PropTypes.string,
    backTextLeft: PropTypes.string,

    buttonTextStyle: PropTypes.any,
    textRightStyle: PropTypes.any,
    textLeftStyle: PropTypes.any,

    buttonStyle: PropTypes.any,
    buttonContainerStyle: PropTypes.any,
    rightContainerStyle: PropTypes.any,
    leftContainerStyle: PropTypes.any,
  };

  componentDidUpdate(prevProps) {
    if (prevProps.switchOn !== this.props.switchOn) {
      this.runAnimation();
    }
  };

  onPress = () => {
    this.props.onPress();
  };

  getStart = () => {
    return this.props.type === undefined ? 0 : this.props.type === 0 ? 0 : this.props.containerStyle.padding * 2;
  };

  runAnimation = () => {
    // this.state.anim.setValue(0);
    const animValue: any = {
      fromValue: this.props.switchOn ? 0 : 1,
      toValue: this.props.switchOn ? 1 : 0,
      duration: this.props.duration,
    };
    Animated.timing(this.state.animXValue, animValue).start();
    // Animated.timing(this.state.anim, animValue).start(() => this.runAnimation());
  };

  constructor(props) {
    super(props);
    const endPos = this.props.containerStyle.width - (this.props.circleStyle.width + this.props.containerStyle.padding * 2);
    this.state = {
      circlePosXStart: this.getStart(),
      circlePosXEnd: endPos,
      animXValue: new Animated.Value(this.props.switchOn ? 1 : 0),
    };
  }

  render() {
    return (
      <TouchableOpacity
        testID='toggle'
        onPress={this.onPress}
        activeOpacity={0.5}
      >
        <Animated.View
          style={[
            styles.container,
            this.props.containerStyle,
            {
              backgroundColor: this.state.animXValue.interpolate({
                inputRange: [ 0, 1 ],
                outputRange: [ this.props.backgroundColorOff, this.props.backgroundColorOn ],
              }),
            },
          ]}
        >
          {
            this.generateLeftText()
          }
          <Animated.View
            style={[
              this.props.circleStyle,
              {
                backgroundColor: this.state.animXValue.interpolate({
                  inputRange: [ 0, 1 ],
                  outputRange: [ this.props.circleColorOff, this.props.circleColorOn ],
                }),
              },
              {
                transform: [
                  {
                    translateX: this.state.animXValue.interpolate({
                      inputRange: [ 0, 1 ],
                      outputRange: [ this.state.circlePosXStart, this.state.circlePosXEnd ],
                    }),
                  },
                ],
              },
              this.props.buttonStyle,
            ]}>
            <Animated.View style={this.props.buttonContainerStyle}>
              <Text style={this.props.buttonTextStyle}>{this.props.buttonText}</Text>
            </Animated.View>
          </Animated.View>
          {
            this.generateRightText()
          }
        </Animated.View>
      </TouchableOpacity>
    );
  }

  generateRightText() {
    return (
      <Animated.View style={this.props.rightContainerStyle}>
        <Text style={this.props.textRightStyle}>{this.props.backTextRight}</Text>
      </Animated.View>
    );
  }

  generateLeftText() {
    return (
      <Animated.View style={this.props.leftContainerStyle}>
        <Text style={this.props.textLeftStyle}>{this.props.backTextLeft}</Text>
      </Animated.View>
    );
  }
}

const styles: any = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignItems: 'center',
  },
});

export default Item;


Example/App.js
/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 */
import React, { Component } from 'react';
import {
  Platform,
  StyleSheet,
  Text,
  View,
} from 'react-native';

import SwitchToggle from 'react-native-switch-toggle';

export default class App extends Component<{}> {
  constructor(props) {
    super(props);
    this.state = {
      switchOn1: false,
      switchOn2: false,
      switchOn4: false,
    };
  }

  getButtonText() {
    return this.state.switchOn4 ? 'Hour' : 'Day';
  }

  getRightText() {
    return this.state.switchOn4 ? '' : 'Hour';
  }

  getLeftText() {
    return this.state.switchOn4 ? 'Day' : '';
  }

  render() {
    return (
      <View style={styles.container}>
        <SwitchToggle
          switchOn={this.state.switchOn1}
          onPress={this.onPress1}
        />
        <SwitchToggle
          type={0}
          containerStyle={{
            marginTop: 16,
            width: 108,
            height: 48,
            borderRadius: 25,
            backgroundColor: '#ccc',
            padding: 5,
          }}
          circleStyle={{
            width: 38,
            height: 38,
            borderRadius: 19,
            backgroundColor: 'white', // rgb(102,134,205)
          }}
          switchOn={this.state.switchOn2}
          onPress={this.onPress2}
          circleColorOff='white'
          circleColorOn='red'
          duration={500}
        />
        <SwitchToggle
          type={0}
          containerStyle={{
            marginTop: 16,
            width: 160,
            height: 65,
            borderRadius: 30,
            padding: 5,
          }}
          backgroundColorOn='#a0e1e5'
          backgroundColorOff='#e5e1e0'
          circleStyle={{
            width: 55,
            height: 55,
            borderRadius: 27.5,
            backgroundColor: 'blue', // rgb(102,134,205)
          }}
          switchOn={this.state.switchOn3}
          onPress={this.onPress3}
          circleColorOff='#ff11ff'
          circleColorOn='green'
          duration={500}
        />
        <SwitchToggle
          buttonText={this.getButtonText()}
          backTextRight={this.getRightText()}
          backTextLeft={this.getLeftText()}
          type={1}
          buttonStyle={{
            alignItems: 'center',
            justifyContent: 'center',
            position: 'absolute',
          }}
          rightContainerStyle={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}
          leftContainerStyle={{ flex: 1, alignItems: 'center', justifyContent: 'flex-start' }}
          buttonTextStyle={{ fontSize: 20 }}
          textRightStyle={{ fontSize: 20 }}
          textLeftStyle={{ fontSize: 20 }}
          containerStyle={{
            marginTop: 16,
            width: 160,
            height: 65,
            borderRadius: 30,
            padding: 5,
          }}
          backgroundColorOn='#fff'
          backgroundColorOff='#fff'
          circleStyle={{
            width: 80,
            height: 55,
            borderRadius: 27.5,
            backgroundColor: 'blue', // rgb(102,134,205)
          }}
          switchOn={this.state.switchOn4}
          onPress={this.onPress4}
          circleColorOff='#e5e1e0'
          circleColorOn='#e5e1e0'
          duration={500}
        />
      </View>
    );
  }

    onPress1 = () => {
      this.setState({ switchOn1: !this.state.switchOn1 });
    };
    onPress2 = () => {
      this.setState({ switchOn2: !this.state.switchOn2 });
    };
    onPress3 = () => {
      this.setState({ switchOn3: !this.state.switchOn3 });
    };
    onPress4 = () => {
      this.setState({ switchOn4: !this.state.switchOn4 });
    };
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#F5FCFF',
  },
});
